# Journey by **Pandas**


Today I will learn all about **Pandas**. It is essential to learn for me.

---
## 1 · Introduction

**Pandas** is an open-source data manipulation and analysis library for **Python**. It’s designed to handle, clean, and analyze data in a way that’s both powerful and intuitive. With **Pandas**, It is easy to load data from various sources, transform it, and perform complex operations.

Lets start the journey by importing **Pandas**...

In [1]:
# importing pandas :
import pandas as pd

# importing numpy :
import numpy as np

print("Pandas version :", pd.__version__)
print('NumPy version :', np.__version__)

Pandas version : 2.3.3
NumPy version : 2.0.2


---
## 2 · DataFrame

**Pandas** primarily works with two data structures: **Series** and **DataFrames**. A **DataFrame** is like a spreadsheet, with rows and columns. 

1. Series : a one-dimensional labeled array holding data of any type such as integers, strings, Python objects etc.
2. DataFrame : a two-dimensional data structure that holds data like a two-dimension array or a table with rows and columns.

Here’s how to create a simple DataFrame:

In [2]:
# creating a DataFrame :
# firstly, making a `Dictionary` with `List` values:
data = {
    'Name' : ['Abdullah'],    # value must be a `List` / `Array`.
    'Age'  : [18],
}

# secondly, convarting Dictionary to DataFrame :
df = pd.DataFrame(data)

print(df, '\n')
print(type(df))

       Name  Age
0  Abdullah   18 

<class 'pandas.core.frame.DataFrame'>


Here's `0` is called `Index`. By default, Pandas shows index as integer. like, 0, 1, 2, and so on. But I can change it with `index=` arguments.

---
Another DataFrame with `Index` :

In [3]:
# a dataframe :
data = { 'Name' : [ 'Saem', 'Siam', 'Maruf', 'Atik', 'Mamun' ],
         'Grade': [ 'B', 'A', 'A+', 'B', 'A' ], }
indexes = ['B', 'A', 'A', 'B', 'A']

result = pd.DataFrame(data, index=indexes)

print(result, '\n')

    Name Grade
B   Saem     B
A   Siam     A
A  Maruf    A+
B   Atik     B
A  Mamun     A 



Now the `Index` column is showing my given indexes.

---
### DataFrame to Series

I can make a Series from DataFrame. Lets do it...

In [4]:
# Dataframe to series :
series_data = pd.Series(result['Grade'])

# to make a list of column :
columns = result.columns.to_list()

print(columns)
print(series_data)
print(type(series_data))

['Name', 'Grade']
B     B
A     A
A    A+
B     B
A     A
Name: Grade, dtype: object
<class 'pandas.core.series.Series'>


---
### NumPy to DataFrame

I also can make DataFrame from NumPy array by just adding column names. I can see DataFrame's shape by writing `pd.shape`.

Lets do it...

In [5]:
# numpy to dataframe :
numpy_array = np.array([[0, 3], [2, 6], [4, 12], [5, 15]])
columns_name = ['temp', 'activity']

print(numpy_array, '\n', numpy_array.shape, '\n')

dataframe = pd.DataFrame(data=numpy_array, columns=columns_name)

dataframe

[[ 0  3]
 [ 2  6]
 [ 4 12]
 [ 5 15]] 
 (4, 2) 



,temp,activity
0,0,3
1,2,6
2,4,12
3,5,15



---
## 3 · Series

**Series** is an one-dimensional labeled array holding data of any type such as integers, strings, Python objects etc.

---
Creating a **Series** by passing a list of values .

In [6]:
# creating a series :
series_data = pd.Series([1, 2, 3, 4, 5, 6, 7])


# creating a series of temperature with Index :
series_indexes = pd.Series(
    [24, 26, 25, 26, 27],
    index=['Morning', 'Noon', 'Evening', 'Night', 'MidNight'])

print(series_data)
print(type(series_data), '\n')

print(series_indexes)
print(series_indexes.shape, '\n')

0    1
1    2
2    3
3    4
4    5
5    6
6    7
dtype: int64
<class 'pandas.core.series.Series'> 

Morning     24
Noon        26
Evening     25
Night       26
MidNight    27
dtype: int64
(5,) 



---
### Series to DataFrame

Lets make a DataFrame from our Series Data...

In [7]:
# Series to Dataframe :
temperature_celsius = pd.DataFrame({
    'Temperature' : series_indexes
})

print(temperature_celsius)
print(type(temperature_celsius))

          Temperature
Morning            24
Noon               26
Evening            25
Night              26
MidNight           27
<class 'pandas.core.frame.DataFrame'>


---
### NumPy to Series

Now I will make a Series from NumPy array.

In [8]:
# numpy to series :
numpy_array = np.array([1, 3, 5, 7, 9])

series_data = pd.Series(data=numpy_array)

series_data

0    1
1    3
2    5
3    7
4    9
dtype: int64

---
## 4 · Operations

**Pandas** gives us a lot of operations for data manipulation.

---
### Adding Data

**Pandas** makes it easy to add new data in a DataFrame :


In [9]:
# making a dataframe :
data = { 'Name' : ['Arefin', 'Anowar', 'Inan', 'Rafid', 'Zarif', 'Labib'],
         'Age' : [ 19, 19, 21, 20, 23, 18 ] }

students = pd.DataFrame(data)

print(students, '\n')

# adding data :
students['Skill'] = ['AI', 'ML', 'CV', 'SL', 'NLP', 'EDA']

print(students, '\n')

     Name  Age
0  Arefin   19
1  Anowar   19
2    Inan   21
3   Rafid   20
4   Zarif   23
5   Labib   18 

     Name  Age Skill
0  Arefin   19    AI
1  Anowar   19    ML
2    Inan   21    CV
3   Rafid   20    SL
4   Zarif   23   NLP
5   Labib   18   EDA 



---
Lets add a new column / feature named 'Category' according to age. 

In [10]:
# adding Data ;
students['Category'] = students['Age'].apply(lambda y: 'Young' if y <= 20 else 'Senior')

students

,Name,Age,Skill,Category
0,Arefin,19,AI,Young
1,Anowar,19,ML,Young
2,Inan,21,CV,Senior
3,Rafid,20,SL,Young
4,Zarif,23,NLP,Senior
5,Labib,18,EDA,Young


---
I can make any Column an index. Lets see :

In [11]:
# making column an index :
students.set_index('Skill')

,Name,Age,Category
Skill,,,
AI,Arefin,19,Young
ML,Anowar,19,Young
CV,Inan,21,Senior
SL,Rafid,20,Young
NLP,Zarif,23,Senior
EDA,Labib,18,Young


---
### Selecting Data

I can select any specific column from DataFrame like this :

In [12]:
# selecting data as like Dictionary :
names = students['Name']

print('Names :\n', names, '\n')
print(type(names), '\n')

Names :
 0    Arefin
1    Anowar
2      Inan
3     Rafid
4     Zarif
5     Labib
Name: Name, dtype: object 

<class 'pandas.core.series.Series'> 



In [13]:
# selecting Column as DataFrame's Properties :
students.Name

0    Arefin
1    Anowar
2      Inan
3     Rafid
4     Zarif
5     Labib
Name: Name, dtype: object

In [14]:
# selecting multiple column :
print(students[['Name', 'Age']])

     Name  Age
0  Arefin   19
1  Anowar   19
2    Inan   21
3   Rafid   20
4   Zarif   23
5   Labib   18


In [15]:
# index based selection ;
# first row of dataframe :
students.iloc[[0]]

,Name,Age,Skill,Category
0,Arefin,19,AI,Young


In [16]:
# row first , column second :
students.iloc[:, 0]

0    Arefin
1    Anowar
2      Inan
3     Rafid
4     Zarif
5     Labib
Name: Name, dtype: object

In [17]:
# row first , column second :
students.iloc[:, 1]

0    19
1    19
2    21
3    20
4    23
5    18
Name: Age, dtype: int64

In [18]:
# to select second and third entry :
students.iloc[1:3, 0]

1    Anowar
2      Inan
Name: Name, dtype: object

In [19]:
# it is also possible by passing a list :
students.iloc[[1, 2, 3], 0]

1    Anowar
2      Inan
3     Rafid
Name: Name, dtype: object

In [20]:
# negative number indexing :
students.iloc[-5:]

,Name,Age,Skill,Category
1,Anowar,19,ML,Young
2,Inan,21,CV,Senior
3,Rafid,20,SL,Young
4,Zarif,23,NLP,Senior
5,Labib,18,EDA,Young


---
### Filtering Data

I can filter data based on specific conditions :

In [21]:
# conditional selecting :
df = students[students.Age == 19]
df

,Name,Age,Skill,Category
0,Arefin,19,AI,Young
1,Anowar,19,ML,Young


In [22]:
# filtering data :
teenager = students[students['Age'] <= 20]

print(teenager, '\n')

     Name  Age Skill Category
0  Arefin   19    AI    Young
1  Anowar   19    ML    Young
3   Rafid   20    SL    Young
5   Labib   18   EDA    Young 



---
### Viewing Data

I can view datapoint from a dataframe in many ways.

In [23]:
# first entry of 'Category'
print(students['Category'][0])

Young


In [24]:
# specifying data :
# first 3 rows :
print(students.head(3))

     Name  Age Skill Category
0  Arefin   19    AI    Young
1  Anowar   19    ML    Young
2    Inan   21    CV   Senior


In [25]:
# just 2nd row :
print(students.iloc[[1]])   # integer location

print()
# 3rd row :
print(students.loc[2])

     Name  Age Skill Category
1  Anowar   19    ML    Young

Name          Inan
Age             21
Skill           CV
Category    Senior
Name: 2, dtype: object


In [26]:
# second to forth ;
print(students[1:4], '\n')

     Name  Age Skill Category
1  Anowar   19    ML    Young
2    Inan   21    CV   Senior
3   Rafid   20    SL    Young 



In [27]:
# just a column :
print(students['Skill'])

0     AI
1     ML
2     CV
3     SL
4    NLP
5    EDA
Name: Skill, dtype: object


---
### Grouping Data 

I can also make group by conditions.

In [28]:
# grouping data :
groupd = students.groupby('Category')['Age'].mean()

print('Average of Age according to age category :')
print(groupd)

Average of Age according to age category :
Category
Senior    22.0
Young     19.0
Name: Age, dtype: float64


---
### Sorting Data

I can sort dataframe by `pd.sort_values()`.

In [29]:
# sorting dataframe :
students.sort_values('Age', ascending=False)

,Name,Age,Skill,Category
4,Zarif,23,NLP,Senior
2,Inan,21,CV,Senior
3,Rafid,20,SL,Young
0,Arefin,19,AI,Young
1,Anowar,19,ML,Young
5,Labib,18,EDA,Young


---
### Describing Data

I can describe DataFrame by `pd.info()` and `pd.describe`.

In [30]:
# information :

students.info()

print()

# by default it only shows numeric fields :
students.describe()

print()

# to see all fields :
students.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Name      6 non-null      object
 1   Age       6 non-null      int64 
 2   Skill     6 non-null      object
 3   Category  6 non-null      object
dtypes: int64(1), object(3)
memory usage: 324.0+ bytes




,Name,Age,Skill,Category
count,6,6.000000,6,6
unique,6,NaN,6,2
top,Arefin,NaN,AI,Young
freq,1,NaN,1,4
mean,NaN,20.000000,NaN,NaN
std,NaN,1.788854,NaN,NaN
min,NaN,18.000000,NaN,NaN
25%,NaN,19.000000,NaN,NaN
50%,NaN,19.500000,NaN,NaN
75%,NaN,20.750000,NaN,NaN


---
## 5 · Data Loading

Now I will learn how to load 'CSV' dataset by **Pandas** and how to make a 'CSV' dataset by it.

---
### Loading CSV

Lets load a csv file from my os.

In [31]:
# loading dataset : i can load the set and make its id to index ; index_col=
#train = pd.read_csv('/kaggle/input/competitions/ghouls-goblins-and-ghosts-boo', index_col=0)

train = pd.read_csv('/kaggle/input/datasets/ibnulmisbah/ghouls-goblins-and-ghosts-boo/train.csv')
test = pd.read_csv('/kaggle/input/datasets/ibnulmisbah/ghouls-goblins-and-ghosts-boo/test.csv')

print('Data Loaded Successfully')

Data Loaded Successfully


This is train Dataset of a Kaggle PlayGround Competition named 'Ghouls, Goblins and Ghosts... Boo!'.

---
### Making CSV

Lets make a csv file from `students` DataFrame. Just using `pd.to_csv` we can make a csv file from a DataFrame.

In [32]:
# making csv from student dataframe :
students.to_csv('InsigniaStudents.csv')

students

,Name,Age,Skill,Category
0,Arefin,19,AI,Young
1,Anowar,19,ML,Young
2,Inan,21,CV,Senior
3,Rafid,20,SL,Young
4,Zarif,23,NLP,Senior
5,Labib,18,EDA,Young


---
## 6 · Real EDA

Now I will do some real Exploratory Data Analysis by **Pandas** with this big CSV file. In the above part I practice **Pandas** with some handy DataFrame and that is also not organized. Now I will do EDA in a organized way.


---
### Exploring Data

Firstly I will explore whole data to see its beauty or trash. Then I will take action what will be best.

---
#### pd.head()

Lets start EDA by exploring the Dataset by `pd.head()` and `pd.tail()`:

In [33]:
# exploring data ;
train

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
0,0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
1,1,0.575560,0.425868,0.531401,0.439899,green,Goblin
2,2,0.467875,0.354330,0.811616,0.791225,black,Ghoul
3,4,0.776652,0.508723,0.636766,0.884464,black,Ghoul
4,5,0.566117,0.875862,0.418594,0.636438,green,Ghost
...,...,...,...,...,...,...,...
366,886,0.458132,0.391760,0.660590,0.635689,blue,Goblin
367,889,0.331936,0.564836,0.539216,0.551471,green,Ghost
368,890,0.481640,0.501147,0.496446,0.544003,clear,Ghoul
369,896,0.294943,0.771286,0.583503,0.300618,clear,Ghost


It is also more easy to see a dataset without no methods and attributes. Just write its name and run it. It will also show its `shape`.

---
#### pd.shape

To see any dataset's shape I use this attriute : `np.shape`.

In [34]:
# shape of the dataset :
train.shape

(371, 7)

It means this dataset contains 371 rows with 7 columns.

---
#### pd.info()

Lets see its informations :

In [35]:
# shoing information :
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371 entries, 0 to 370
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             371 non-null    int64  
 1   bone_length    371 non-null    float64
 2   rotting_flesh  371 non-null    float64
 3   hair_length    371 non-null    float64
 4   has_soul       371 non-null    float64
 5   color          371 non-null    object 
 6   type           371 non-null    object 
dtypes: float64(4), int64(1), object(2)
memory usage: 20.4+ KB


It is very good news that it has no missing value entirely.

---
#### pd.describe()

Lets describe it :

In [36]:
# describing the set :
train.describe()

,id,bone_length,rotting_flesh,hair_length,has_soul
count,371.000000,371.000000,371.000000,371.000000,371.000000
mean,443.676550,0.434160,0.506848,0.529114,0.471392
std,263.222489,0.132833,0.146358,0.169902,0.176129
min,0.000000,0.061032,0.095687,0.134600,0.009402
25%,205.500000,0.340006,0.414812,0.407428,0.348002
50%,458.000000,0.434891,0.501552,0.538642,0.466372
75%,678.500000,0.517223,0.603977,0.647244,0.600610
max,897.000000,0.817001,0.932466,1.000000,0.935721


It is only describes numerical features. but I can see all features by it . Lets see it :

In [37]:
# describing all features :
train.describe(include='all')

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
count,371.000000,371.000000,371.000000,371.000000,371.000000,371,371
unique,NaN,NaN,NaN,NaN,NaN,6,3
top,NaN,NaN,NaN,NaN,NaN,white,Ghoul
freq,NaN,NaN,NaN,NaN,NaN,137,129
mean,443.676550,0.434160,0.506848,0.529114,0.471392,NaN,NaN
std,263.222489,0.132833,0.146358,0.169902,0.176129,NaN,NaN
min,0.000000,0.061032,0.095687,0.134600,0.009402,NaN,NaN
25%,205.500000,0.340006,0.414812,0.407428,0.348002,NaN,NaN
50%,458.000000,0.434891,0.501552,0.538642,0.466372,NaN,NaN
75%,678.500000,0.517223,0.603977,0.647244,0.600610,NaN,NaN


---
So, I have looked in this dataset in some ways and I found that :

- It is not missing any value.
- It has 5 numerical features with `id`.
- It has 2 categorical features with target feature.
- Target (`type`) feature has only 3 unique values.
- `Color` feature has 6 unique values.

Now I will see very attantively the categorical features...

In [38]:
# color column with its value :
train.color

0      clear
1      green
2      black
3      black
4      green
       ...  
366     blue
367    green
368    clear
369    clear
370    white
Name: color, Length: 371, dtype: object

---
#### pd.mean()

Lets see dataset's `hair_length`'s mean :

In [39]:
train.hair_length.mean()

np.float64(0.5291143100058726)

---
#### pd.median()

Lets see dataset's `hair_length`'s median :

In [40]:
train.hair_length.median()

0.5386424792099368

---
#### pd.isin()

Lets see black and white color's set. Pandas comes with a few built-in conditional selectors, on of them  is `isin`. `isin` is lets us select data whose value "is in" a list of values. For example :

In [41]:
train.color.isin(['black', 'white'])

0      False
1      False
2       True
3       True
4      False
       ...  
366    False
367    False
368    False
369    False
370     True
Name: color, Length: 371, dtype: bool

---
#### pd.isna()

If in this dataset any missing value exixted, I can find them by  `pd.isna()` or `pd.isnull()`. These two function are same.

In [42]:
# finding missing data :
train.isna()

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
0,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...
366,False,False,False,False,False,False,False
367,False,False,False,False,False,False,False
368,False,False,False,False,False,False,False
369,False,False,False,False,False,False,False


---
#### pd.notna()

To see which values are existed I can use `pd.notna()` or `pd.notnull()`.

In [43]:
# finding non-null data :
train.notna()

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
0,True,True,True,True,True,True,True
1,True,True,True,True,True,True,True
2,True,True,True,True,True,True,True
3,True,True,True,True,True,True,True
4,True,True,True,True,True,True,True
...,...,...,...,...,...,...,...
366,True,True,True,True,True,True,True
367,True,True,True,True,True,True,True
368,True,True,True,True,True,True,True
369,True,True,True,True,True,True,True


---
#### pd.value_counts()

To count values from a column I can use `pd.value_counts()` . Lets see our `type's` value_counts :

In [44]:
# colors value counts :
train.type.value_counts()

type
Ghoul     129
Goblin    125
Ghost     117
Name: count, dtype: int64

In [45]:
# 'black' and `white` color value counts :
train.color.isin(['black', 'white']).value_counts()

color
False    193
True     178
Name: count, dtype: int64

---
#### pd.column.unique()

In this output I found 5 color but the `info` shows it has 6 unique values.

Lets see its unique values :

In [46]:
# unique values :
train.color.unique()

array(['clear', 'green', 'black', 'white', 'blue', 'blood'], dtype=object)

---
Lets see Target features `unique` values .

In [47]:
# target features :
train.type.unique()

array(['Ghoul', 'Goblin', 'Ghost'], dtype=object)

---
### Selecting Data

Now I will select data , make new DataFrame and manipulate tha dataset by many ways .

Firstly I will see the `type` column which is Target feature, very well.

In [48]:
# target feature :
train.type

0       Ghoul
1      Goblin
2       Ghoul
3       Ghoul
4       Ghost
        ...  
366    Goblin
367     Ghost
368     Ghoul
369     Ghost
370     Ghoul
Name: type, Length: 371, dtype: object

---
#### Dictionary Selection

Lets see `color` and `type` columns :

In [49]:
train[['type', 'color']]

,type,color
0,Ghoul,clear
1,Goblin,green
2,Ghoul,black
3,Ghoul,black
4,Ghost,green
...,...,...
366,Goblin,blue
367,Ghost,green
368,Ghoul,clear
369,Ghost,clear


---
#### Index-Based Selection

There is tow common attribute to select DataPoint, one is `pd.loc()` which is label based Data Selection attribute and other is `pd.iloc` which is only index based Data Selection attribute. 

* `pd.loc[row, column]`
* `pd.iloc[row, column]`

`iloc` uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So `0:10` will select entries `0,...,9`. loc, meanwhile, indexes inclusively. So `0:10` will select entries `0,...,10`.

Lets see some example :

In [50]:
# 50th row :
train.loc[50]

id                    103
bone_length      0.571619
rotting_flesh    0.431955
hair_length      0.641309
has_soul         0.462295
color               black
type                Ghoul
Name: 50, dtype: object

In [51]:
# more visually :
train.loc[[50]]

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
50,103,0.571619,0.431955,0.641309,0.462295,black,Ghoul


I think `Ghoul` is more than others. Lets see a set of ghouls.

In [52]:
# ghoul dataset 
train.loc[train.type == 'Ghoul']

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
0,0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
2,2,0.467875,0.354330,0.811616,0.791225,black,Ghoul
3,4,0.776652,0.508723,0.636766,0.884464,black,Ghoul
7,11,0.516224,0.536429,0.612776,0.468048,clear,Ghoul
11,23,0.584543,0.593082,0.681166,0.935721,clear,Ghoul
...,...,...,...,...,...,...,...
360,864,0.467785,0.420135,0.631894,0.386509,clear,Ghoul
361,867,0.501022,0.584847,0.656230,0.469886,clear,Ghoul
365,878,0.406062,0.449944,0.788983,0.428553,white,Ghoul
368,890,0.481640,0.501147,0.496446,0.544003,clear,Ghoul


It seems Ghoul color is clear for many of this .

In [53]:
# color dataset with clear :
train.loc[train.color == 'clear']

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
0,0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
7,11,0.516224,0.536429,0.612776,0.468048,clear,Ghoul
10,22,0.431685,0.438959,0.239212,0.471820,clear,Goblin
11,23,0.584543,0.593082,0.681166,0.935721,clear,Ghoul
14,28,0.513387,0.301345,0.745676,0.545792,clear,Goblin
...,...,...,...,...,...,...,...
361,867,0.501022,0.584847,0.656230,0.469886,clear,Ghoul
362,873,0.377449,0.466158,0.497351,0.393305,clear,Goblin
364,877,0.486744,0.462666,0.840013,0.353383,clear,Goblin
368,890,0.481640,0.501147,0.496446,0.544003,clear,Ghoul


---
Now I will see Ghost dataset :

In [54]:
# all ghost :
ghost_set = train.loc[train.type == 'Ghost']
ghost_set

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
4,5,0.566117,0.875862,0.418594,0.636438,green,Ghost
8,12,0.314295,0.671280,0.417267,0.227548,blue,Ghost
9,19,0.280942,0.701457,0.179633,0.141183,white,Ghost
16,30,0.250770,0.246258,0.554654,0.250036,black,Ghost
19,34,0.524080,0.750988,0.524637,0.443358,green,Ghost
...,...,...,...,...,...,...,...
354,847,0.356999,0.671331,0.296784,0.689980,black,Ghost
358,858,0.440432,0.309498,0.203057,0.194105,clear,Ghost
363,874,0.507583,0.799623,0.334520,0.344403,white,Ghost
367,889,0.331936,0.564836,0.539216,0.551471,green,Ghost


---
Lets see all about ghost dataset :

In [55]:
# all about ghost dataset :
ghost_set.describe(include='all')

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
count,117.000000,117.000000,117.000000,117.000000,117.000000,117,117
unique,NaN,NaN,NaN,NaN,NaN,6,1
top,NaN,NaN,NaN,NaN,NaN,white,Ghost
freq,NaN,NaN,NaN,NaN,NaN,44,117
mean,436.538462,0.347466,0.593759,0.373186,0.320023,NaN,NaN
std,250.974945,0.117748,0.142150,0.117642,0.139790,NaN,NaN
min,5.000000,0.061032,0.165539,0.134600,0.009402,NaN,NaN
25%,222.000000,0.269748,0.504399,0.299507,0.230361,NaN,NaN
50%,428.000000,0.340442,0.603780,0.365142,0.307534,NaN,NaN
75%,641.000000,0.417253,0.682687,0.461068,0.415930,NaN,NaN


In [56]:
# value counts of colors :
ghost_set.color.value_counts()

color
white    44
clear    32
green    15
black    14
blue      6
blood     6
Name: count, dtype: int64

---
Lets make a set of `blood` color :

In [57]:
# blood color selections :
blood_set = train.loc[train.color == 'blood']
blood_set

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
46,93,0.481716,0.527251,0.740270,0.700857,blood,Ghoul
117,282,0.453869,0.259701,0.581400,0.567843,blood,Goblin
151,358,0.332905,0.494895,0.474788,0.514944,blood,Ghost
158,375,0.250523,0.650506,0.384925,0.385882,blood,Ghost
171,412,0.481141,0.696810,0.162611,0.453465,blood,Ghost
271,664,0.184172,0.715131,0.134600,0.227068,blood,Ghost
287,695,0.479818,0.569762,0.721362,0.491938,blood,Ghoul
290,699,0.417253,0.624973,0.328752,0.279668,blood,Ghost
294,708,0.358240,0.370775,0.667384,0.740579,blood,Ghoul
324,772,0.649559,0.507846,0.734855,0.546648,blood,Ghoul


---
Lets see a dataset of blood color and Ghoul type :

In [58]:
train.loc[(train.color == 'blood') & (train.type == 'Ghoul')]

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
46,93,0.481716,0.527251,0.740270,0.700857,blood,Ghoul
287,695,0.479818,0.569762,0.721362,0.491938,blood,Ghoul
294,708,0.358240,0.370775,0.667384,0.740579,blood,Ghoul
324,772,0.649559,0.507846,0.734855,0.546648,blood,Ghoul


---
Lets see a dataset of blood color and Ghost type :

In [59]:
train.loc[(train.color == 'blood') & (train.type == 'Ghost')]

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
151,358,0.332905,0.494895,0.474788,0.514944,blood,Ghost
158,375,0.250523,0.650506,0.384925,0.385882,blood,Ghost
171,412,0.481141,0.696810,0.162611,0.453465,blood,Ghost
271,664,0.184172,0.715131,0.134600,0.227068,blood,Ghost
290,699,0.417253,0.624973,0.328752,0.279668,blood,Ghost
328,785,0.396988,0.611109,0.507214,0.179642,blood,Ghost


---
Lets see a dataset of blood color and Goblin type :

In [60]:
train.loc[(train.color == 'blood') & (train.type == 'Goblin')]

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
117,282,0.453869,0.259701,0.581400,0.567843,blood,Goblin
348,839,0.355313,0.344684,0.555728,0.638232,blood,Goblin


---
Lets see a dataset of blood and blue color :

In [61]:
train.loc[train.color.isin(['blood', 'blue'])]

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
8,12,0.314295,0.671280,0.417267,0.227548,blue,Ghost
18,32,0.605836,0.587943,0.529361,0.526718,blue,Ghoul
35,75,0.456371,0.456208,0.236332,0.300486,blue,Ghost
46,93,0.481716,0.527251,0.740270,0.700857,blood,Ghoul
55,115,0.678519,0.476821,0.477152,0.489358,blue,Goblin
84,189,0.375197,0.742953,0.320764,0.243440,blue,Ghost
99,222,0.313506,0.616803,0.396482,0.447946,blue,Ghost
107,247,0.452741,0.384133,0.305594,0.371807,blue,Goblin
117,282,0.453869,0.259701,0.581400,0.567843,blood,Goblin
134,315,0.358311,0.621728,0.559499,0.761621,blue,Goblin


In [62]:
train.loc[train.color.isin(['blood', 'blue'])].type.value_counts()

type
Ghost     12
Ghoul     10
Goblin     9
Name: count, dtype: int64

---
### Mapping Data

A `map` is a term, borrowed from mathematics, for a function that takes one set of values and "maps" them to another set of values. In data science we often have a need for creating new representations from existing data, or for transforming data from the format it is in now to the format that we want it to be in later. Maps are what handle this work, making them extremely important for getting your work done!

---
There are two mapping methods that I will use often :

#### pd.map()

`map()` is the first, and slightly simpler one. For example, suppose that we wanted to multiply the `has_soul` column by 100 to make percentages, we can use it :

In [63]:
# map() fucntions to maps :
train.has_soul.map(lambda x: x * 100 )

0      78.114167
1      43.989888
2      79.122497
3      88.446369
4      63.643782
         ...    
366    63.568919
367    55.147051
368    54.400323
369    30.061804
370    60.838396
Name: has_soul, Length: 371, dtype: float64

---
But without `map()` I can do this with **Pandas** builtins operations.

In [64]:
# maping without maps :
train.has_soul * 100

0      78.114167
1      43.989888
2      79.122497
3      88.446369
4      63.643782
         ...    
366    63.568919
367    55.147051
368    54.400323
369    30.061804
370    60.838396
Name: has_soul, Length: 371, dtype: float64

But `map()` is more flexible and builtins are more speedy.

---
#### pd.apply()

The `map()` function should expect a single value from the Series and return a transformed version of that value. `map()` returns a new Series where all the values have been transformed by function.

`apply()` is the equivalent method if I want to transform a whole DataFrame by calling a custom method on each row.

Lets see it :

In [65]:
train.has_soul.apply(lambda x: x * 100)

0      78.114167
1      43.989888
2      79.122497
3      88.446369
4      63.643782
         ...    
366    63.568919
367    55.147051
368    54.400323
369    30.061804
370    60.838396
Name: has_soul, Length: 371, dtype: float64

In [66]:
train.idxmax()

id               370
bone_length      166
rotting_flesh     52
hair_length       17
has_soul          11
color              6
type               1
dtype: int64

---

## 7 · Functions

Now I will write some new function which learned from kaggle Pandas Course :

---
### Methods

#### pd.astype()

It's possible to convert a column of one type into another wherever such a conversion makes sense by using the astype() function. For example, we may transform the points column from its existing int64 data type into a float64 data type:

In [67]:
train.has_soul.astype('int64')

0      0
1      0
2      0
3      0
4      0
      ..
366    0
367    0
368    0
369    0
370    0
Name: has_soul, Length: 371, dtype: int64

---
#### pd.fillna()

Replacing missing values is a common operation. Pandas provides a really handy method for this problem: `fillna()`.


In [68]:

train.fillna('Unknown')

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
0,0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
1,1,0.575560,0.425868,0.531401,0.439899,green,Goblin
2,2,0.467875,0.354330,0.811616,0.791225,black,Ghoul
3,4,0.776652,0.508723,0.636766,0.884464,black,Ghoul
4,5,0.566117,0.875862,0.418594,0.636438,green,Ghost
...,...,...,...,...,...,...,...
366,886,0.458132,0.391760,0.660590,0.635689,blue,Goblin
367,889,0.331936,0.564836,0.539216,0.551471,green,Ghost
368,890,0.481640,0.501147,0.496446,0.544003,clear,Ghoul
369,896,0.294943,0.771286,0.583503,0.300618,clear,Ghost


---
#### pd.replace()

Alternatively, we may have a non-null value that we would like to replace. One way to reflect this in the dataset is using the `replace()` method.

In [69]:
train.type.replace('Goblin', 'Ghost')

0      Ghoul
1      Ghost
2      Ghoul
3      Ghoul
4      Ghost
       ...  
366    Ghost
367    Ghost
368    Ghoul
369    Ghost
370    Ghoul
Name: type, Length: 371, dtype: object

----
#### pd.groupby()



In [70]:
train.groupby(['type', 'has_soul']).hair_length.mean()

type    has_soul
Ghost   0.009402    0.321867
        0.036457    0.462629
        0.043321    0.218732
        0.055464    0.313948
        0.064468    0.235669
                      ...   
Goblin  0.698965    0.471120
        0.708981    0.378707
        0.756766    0.448989
        0.761621    0.559499
        0.854949    0.756629
Name: hair_length, Length: 371, dtype: float64

---
#### pd.sort_values()

To get data in the order want it in we can sort it ourselves. The `sort_values()` method is handy for this.

In [71]:
train.bone_length.sort_values(ascending=False)

166    0.817001
3      0.776652
284    0.761355
304    0.757138
249    0.755298
         ...   
344    0.165196
251    0.160465
175    0.145071
177    0.100945
174    0.061032
Name: bone_length, Length: 371, dtype: float64

---
#### pd.sort_index()

To sort by index values, use the companion method `sort_index()`. This method has the same arguments and default order:

In [72]:
train.bone_length.sort_index()

0      0.354512
1      0.575560
2      0.467875
3      0.776652
4      0.566117
         ...   
366    0.458132
367    0.331936
368    0.481640
369    0.294943
370    0.670200
Name: bone_length, Length: 371, dtype: float64

---
#### pd.rename()

`rename()` lets us change index names and/or column names. For example :

In [73]:
# renaming column `has_soul` to `soul`
train.rename(columns={'has_soul': 'soul'})

,id,bone_length,rotting_flesh,hair_length,soul,color,type
0,0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
1,1,0.575560,0.425868,0.531401,0.439899,green,Goblin
2,2,0.467875,0.354330,0.811616,0.791225,black,Ghoul
3,4,0.776652,0.508723,0.636766,0.884464,black,Ghoul
4,5,0.566117,0.875862,0.418594,0.636438,green,Ghost
...,...,...,...,...,...,...,...
366,886,0.458132,0.391760,0.660590,0.635689,blue,Goblin
367,889,0.331936,0.564836,0.539216,0.551471,green,Ghost
368,890,0.481640,0.501147,0.496446,0.544003,clear,Ghoul
369,896,0.294943,0.771286,0.583503,0.300618,clear,Ghost


In [74]:
# renaming index 1 and 2 :
train.rename(index={0 : 'First', 1 : 'Second'})

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
First,0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
Second,1,0.575560,0.425868,0.531401,0.439899,green,Goblin
2,2,0.467875,0.354330,0.811616,0.791225,black,Ghoul
3,4,0.776652,0.508723,0.636766,0.884464,black,Ghoul
4,5,0.566117,0.875862,0.418594,0.636438,green,Ghost
...,...,...,...,...,...,...,...
366,886,0.458132,0.391760,0.660590,0.635689,blue,Goblin
367,889,0.331936,0.564836,0.539216,0.551471,green,Ghost
368,890,0.481640,0.501147,0.496446,0.544003,clear,Ghoul
369,896,0.294943,0.771286,0.583503,0.300618,clear,Ghost


---
#### pd.set_index()

We'll probably rename columns very often, but rename index values very rarely. For that, `set_index()` is usually more convenient.

In [75]:
train.set_index('id')

,bone_length,rotting_flesh,hair_length,has_soul,color,type
id,,,,,,
0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
1,0.575560,0.425868,0.531401,0.439899,green,Goblin
2,0.467875,0.354330,0.811616,0.791225,black,Ghoul
4,0.776652,0.508723,0.636766,0.884464,black,Ghoul
5,0.566117,0.875862,0.418594,0.636438,green,Ghost
...,...,...,...,...,...,...
886,0.458132,0.391760,0.660590,0.635689,blue,Goblin
889,0.331936,0.564836,0.539216,0.551471,green,Ghost
890,0.481640,0.501147,0.496446,0.544003,clear,Ghoul


---
#### pd.rename_axis()

Both the row index and the column index can have their own name attribute. The complimentary `rename_axis()` method may be used to change these names. For example:

In [76]:
# changing indexes' name attribute :
train.rename_axis(index='Index No', columns='Features')

Features,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
Index No,,,,,,,
0,0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
1,1,0.575560,0.425868,0.531401,0.439899,green,Goblin
2,2,0.467875,0.354330,0.811616,0.791225,black,Ghoul
3,4,0.776652,0.508723,0.636766,0.884464,black,Ghoul
4,5,0.566117,0.875862,0.418594,0.636438,green,Ghost
...,...,...,...,...,...,...,...
366,886,0.458132,0.391760,0.660590,0.635689,blue,Goblin
367,889,0.331936,0.564836,0.539216,0.551471,green,Ghost
368,890,0.481640,0.501147,0.496446,0.544003,clear,Ghoul


---
#### pd.concat()

Sometimes need to combine different DataFrames and/or Series in non-trivial ways. The simplest combining method is `concat()`. Given a list of elements, this function will smush those elements together along an axis.

In [77]:
# Combining two dataset :
pd.concat([train, test])

,id,bone_length,rotting_flesh,hair_length,has_soul,color,type
0,0,0.354512,0.350839,0.465761,0.781142,clear,Ghoul
1,1,0.575560,0.425868,0.531401,0.439899,green,Goblin
2,2,0.467875,0.354330,0.811616,0.791225,black,Ghoul
3,4,0.776652,0.508723,0.636766,0.884464,black,Ghoul
4,5,0.566117,0.875862,0.418594,0.636438,green,Ghost
...,...,...,...,...,...,...,...
524,893,0.377573,0.390158,0.696465,0.355373,blue,NaN
525,894,0.229161,0.601265,0.191282,0.475115,clear,NaN
526,895,0.510497,0.498347,0.708020,0.714154,white,NaN
527,898,0.331472,0.765835,0.338207,0.193431,clear,NaN


---
### Attributes

#### pd.dtype

You can use the `dtype` property to grab the type of a specific column.

In [78]:
train.dtypes

id                 int64
bone_length      float64
rotting_flesh    float64
hair_length      float64
has_soul         float64
color             object
type              object
dtype: object

----
# Conclusion

Finally, I have finished the **Pandas** course from **Kaggle**. I think, I have learned some basic now which will be perfect for me to continue in my **Machine Learning** journey.